portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -87,8 +88,14 @@ public void doGet(HttpServletRequest request, HttpServletResponse response) thro
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
+    int numQueries = getLimit(request);
+    List<Entity> limitedResults = results.asList(FetchOptions.Builder.withLimit(10));
+    if(numQueries != -1){
+        limitedResults = results.asList(FetchOptions.Builder.withLimit(numQueries));
+    }
```

Comment:
If i understand your goal here correctly, you want to fetch the limit sent by the request and, if there is a valid option, use that value to get results. otherwise, default to 10. there are a few things we could change here that would make for better code:

- the way this is right now, you'll sometimes make two calls to results.asList. We can ensure we only ever make one call by rearranging things. If we get rid of the first assignment for limitedResults and move it to an else case below the if, we'll only ever make the call once. However, we're now rewriting essentially the same line twice, just with a different param. So what if, instead, we changed the logic that happens in our if/else to change the value of numQueries? Something like if (numQueries == -1) {numQueries = 10;} and then call results.asList afterwards with the numQueries param.

- the default value of 10 is not clear. This is what we call a "magic number" because we don't document this anywhere. It's often a good idea to name constants. that is, we can introduce a variable like "default_limit", assign it to 10, and then use that variable in the code rather than directly using the number 10

- Defaulting to 10 is fine but the way we do it could use some work. If we have a value of -1, we want to change that value to 10. but why -1? the getLimit function is lacking documentation that says we return -1 on error so someone calling the function would have no idea that's an option without reading code. So we can do one of a few things. We can add documentation and call it a day.  We can not return -1 and let the caller handle negative values. Or, we can throw an error when we get an unexpected value (and document that we throw this error). That way, if a value is truly incorrect, we force the caller to handle getting bad numbers. This prevents the caller from making mistakes as errors need to be handled or else the program will crash. So if this function throws an error (like IllegalArgumentException) and the caller needs to catch that error if it happens and handle accordingly (like using a default value).

Reply:
The first 2 bullet points I addressed. For the third one, though, I did a few things. Firstly, I added a line of documentation saying that if it returns -1 its an error. Secondly, I changed the main function to handle negatives and extremely large numbers. Is that enough or does my approach need more work?

Reply:
nope, that looks good to me! It's also not a bad idea to let it throw that error when it happens instead of returning -1 and just let the caller catch the error, but the fact that -1 is documented is fine as well. Some folks might argue strongly one way or another but I think it's fine this way for now and we can revisit later if we need to.